OpenRA.Mods.Common/Traits/GivesExperience.cs

```diff
@@ -32,31 +33,36 @@ class GivesExperienceInfo : TraitInfo
 		public override object Create(ActorInitializer init) { return new GivesExperience(init.Self, this); }
 	}
 
-	class GivesExperience : INotifyKilled
+	class GivesExperience : INotifyKilled, INotifyCreated
 	{
 		readonly GivesExperienceInfo info;
 
+		int exp;
+		IEnumerable<int> experienceModifiers;
+
 		public GivesExperience(Actor self, GivesExperienceInfo info)
 		{
 			this.info = info;
 		}
 
+		void INotifyCreated.Created(Actor self)
+		{
+			var valued = self.Info.TraitInfoOrDefault<ValuedInfo>();
+			exp = info.Experience >= 0 ? info.Experience
+				: valued != null ? valued.Cost : 0;
+
+			experienceModifiers = self.TraitsImplementing<IGivesExperienceModifier>().ToArray().Select(m => m.GetGivesExperienceModifier());
```

Comment:
You prolly want to have the `ToArray()` at the end.

Reply:
Actually, I don't think we want that. You want to cache the result of `TraitsImplementing` but not the modifiers, since those can change over time.

Reply:
Yes, we use the same approach in `Armament` for the same reason.